[JIT] add CondValue to unify refinements and code emission

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26572 [jit] Allow 'Any'  to appear as a type argument.
* #26499 [jit] refactor tryMatchSchema
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* #26269 [jit] add dynamic isinstance
* #26268 Allow types as node attributes
* **#26145 [JIT] add CondValue to unify refinements and code emission**

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

Differential Revision: [D17359500](https://our.internmc.facebook.com/intern/diff/D17359500)

zdevito committed:
[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -36,115 +36,153 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
+    // so we take their intersection.
+    return RefinementSet(
+        unionList(true_refinements_, rhs.true_refinements_),
+        intersectList(false_refinements_, rhs.false_refinements_));
```
`intersectList` but the struct is named `refinementSet`

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -36,115 +36,153 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
-        ret.setRefinement(name_mapping.first, name_mapping.second);
+  static Refinements intersectList(const Refinements& a, const Refinements& b) {
+    Refinements result;
+    for (const Refinement& r : a) {
```
What is the motivation is for removing the existing `intersection ` and `union` implementations? The new implementation probably suffices for just optional refinement, but with the possibility of more complicated refinement using `isinstance` and an `Any` type i think we should use the more correct existing implementation. 

Something like `isinstance(x, float) or isinstance(x, int)` would just accept the second type instead of unifying them.  

Note: the existing `intersection` and `union` produces the correct refinements, just the `insertRefinements` function incorrectly inserts a refinement when the refinement is not None.

zdevito replied:
discussed in person: there may be unexpected consequences of unifyTypes refinement. For instance if we derive a type Scalar, it is another instance of injecting a Scalar in addition to the item call. This has led to surprising bugs in the past, so we should have a good example case before we make more liberal use of refinement. The behavior currently implemented is a subset of the valid refinements that we would do in the future, so it is safe to expand.

eellison reviewed (commented):
Cool!  `CondValue` will lead to less surprising user behavior around selective compilation of if branches. 

 I have a couple questions / comments on some decisions made. Also i don't think we should regress the refinement calculation for `intersection` and `union` given that we are opening the door for more complicated refinements with `isinstance` checks moving forward.

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -36,115 +36,153 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
+  // with its true value and its false value. If a boolean that has refinements
+  // associated with it is used in a conditional of an if statememt, the true
+  // and false refinements are inserted into the corresponding blocks
+  using Refinements = std::vector<Refinement>;
```
fixed the union as we discussed in person, it now behaves more like a set.

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -899,13 +932,127 @@ struct to_ir {
+    // Refinement none(name, RefinementKind::OPTIONAL_NONE);
+
+    Refinement present(
+        name, lhs_value->type()->expect<OptionalType>()->getElementType());
```
Probably don't want to `expect<OptionalType>` here because it will error for something like:
```
a: int = 1
if a is None:
```

which can happen with a constant attribute that is initialized non-null

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -36,115 +36,148 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
+    return a.identifier() == b.identifier();
+  }
+  static Refinements unionList(const Refinements& a, const Refinements& b) {
+    Refinements result = a;
```
Could consider calling this `unionSet` now that we are enforcing uniqueness of each element

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -36,115 +36,148 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
-        ret.setRefinement(name_mapping.first, name_mapping.second);
+    return result;
+  }
+  static Refinements intersectList(const Refinements& a, const Refinements& b) {
```
Same comment

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1086,38 +1218,57 @@ struct to_ir {
+    // submodule is absent, an is None check can be used to ensure the
+    // accesses to the None check, which would error, are not compiled.
+    if (cond_value.staticIf()) {
+      if (*cond_value.staticIf()) {
```
Not compiling one branch of an if statement with multiple `staticIf` checks is a new feature so maybe add tests for that

eellison reviewed (approved):
Looks good!

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head

zdevito committed:
Update on "[JIT] add CondValue to unify refinements and code emission"

[JIT] add CondValue to unify refinements and code emission

This is step towards isinstance type refinement.
It primarily does yak shaving in compiler.cpp to unify the handling
of special case behavior that occurs in conditional expressions:

* Handling type refinement as part of emission.
* Handling `is None` static-if specialization.

It introduces a CondValue object that is a Value that also has
additional type refinements that are true when that Value is true,
and potentialy a static-true/false value that, if set, will cause if
statements to be handled statically, omitting typechecking of the other side.

This ends up expanding some behavior, for instance `is None` specialization
used to happen only for single expressions, but now works through
boolean logic.

gh-metadata: pytorch pytorch 26145 gh/zdevito/107/head
